### ISSUE TITLE
sol2 tests fail when Lua is compiled as C++

### ISSUE THREAD

**Issue by RaptorFactor** [ThePhD/sol2#239]
Problem: Sol2 tests fail when Lua is compiled as C++ (not sure what else breaks, I hadn't got that far yet).

sol2 version: v2.14.9
Lua version: v5.3.3
Platform: Windows
Compiler: MSVC v19.00.24215.1 x64 (GCC v6.2.0 on mingw-w64 also suspected, although I haven't had time to confirm that yet)

This is what I hacked together to build Lua for this test (/TP and /EHsc are the most important things I think... I use a DLL here but iirc it should repro with a static lib also):
`setlocal`
`rd bin & md bin & cd bin`
`set LUA_SOURCES=..\src\lapi.c ..\src\lauxlib.c ..\src\lbaselib.c ..\src\lbitlib.c ..\src\lcode.c ..\src\lcorolib.c ..\src\lctype.c ..\src\ldblib.c ..\src\ldebug.c ..\src\ldo.c ..\src\ldump.c ..\src\lfunc.c ..\src\lgc.c ..\src\linit.c ..\src\liolib.c ..\src\llex.c ..\src\lmathlib.c ..\src\lmem.c ..\src\loadlib.c ..\src\lobject.c ..\src\lopcodes.c ..\src\loslib.c ..\src\lparser.c ..\src\lstate.c ..\src\lstring.c ..\src\lstrlib.c ..\src\ltable.c ..\src\ltablib.c ..\src\ltm.c ..\src\lundump.c ..\src\lutf8lib.c ..\src\lvm.c ..\src\lzio.c`
`cl /EHsc /MDd /TP /LD /DLUA_BUILD_AS_DLL /Felua53.dll %LUA_SOURCES%`
`cl /EHsc /MDd /TP /DLUA_BUILD_AS_DLL ..\src\lua.c /link lua53.lib`
`cl /EHsc /MDd /TP ..\src\luac.c %LUA_SOURCES%`

Note: In order to compile Lua as C++ you need to make a couple of tweaks to some code that is legal in C but illegal in C++, but it's only a couple of lines and it's as simple as adding a cast. You'll also need to modify lua.hpp to remove the extern "C" wrapper. More on that below...

Here's the one for Sol2:
`del tests.exe`
`del *.obj`
`cl /MDd /EHsc /bigobj /Isingle/sol /Icatch/include /IC:\My\Code\lua-5.3.3\src tests.cpp test_containers.cpp test_coroutines.cpp test_customizations.cpp test_functions.cpp test_inheritance.cpp test_operators.cpp test_overflow.cpp test_simple_usertypes.cpp test_strings.cpp test_tables.cpp test_usertypes.cpp /link C:\My\Code\lua-5.3.3\bin\lua53.lib `

Expected test summary output:

> All tests passed (940 assertions in 138 test cases)

Actual test summary output:
> test cases: 138 | 137 passed | 1 failed
> assertions: 930 | 929 passed | 1 failed

Failing test output:

> -------------------------------------------------------------------------------
> functions/function_result-protected_function_result
> -------------------------------------------------------------------------------
> test_functions.cpp(269)
> ...............................................................................
> 
> test_functions.cpp(269): FAILED:
>   {Unknown expression after the reported line}
> due to unexpected exception with message:
>   lua: error: stack index 1, expected string, received function

Back to compiling Lua as C++... I was under the impression that this was supposed to be a supported and tested scenario, because the Lua website states "Lua also compiles cleanly as C++.", but I guess nobody is really using it in practice? I'm curious why this is... I'm just on the initial bring-up stages of my new project which uses Lua, but I thought that in order to be able to throw exceptions through the Lua DLL that I'd need to compile it as C++, which in turn led me to assume there were a lot of other people who must be doing the same thing... Is this incorrect? How is everyone else dealing with exceptions? (e.g. A.exe -> Lua53.dll -> B.dll - Assuming the methods exposed to Lua from B.dll can throw exceptions)

Also, what's up with lua.hpp? It says that extern "C" is not automatically provided (presumably because people like me want to compile Lua as C++), and then they proceed to do exactly the thing they just said they wouldn't do... I feel like the fact that I'm hitting all these problems means I must be "holding it wrong", so I'd be very curious to hear if you think I'm on the wrong path for whatever reason.

P.S. Sorry, I know you're not responsible for Lua itself, but I haven't been able to find a good answer anywhere else, and now that I already have your attention I may as well ask. ;)

**Comment by griwes** [comment]
*Why* are you trying to do this?

**Comment by RaptorFactor** [comment]
>I'm just on the initial bring-up stages of my new project which uses Lua, but I thought that in order to be able to throw exceptions through the Lua DLL that I'd need to compile it as C++, which in turn led me to assume there were a lot of other people who must be doing the same thing... Is this incorrect? How is everyone else dealing with exceptions? (e.g. A.exe -> Lua53.dll -> B.dll - Assuming the methods exposed to Lua from B.dll can throw exceptions)

**Comment by milleniumbug** [comment]
Unsupported configurations are unsupported.

**Comment by RaptorFactor** [comment]
https://www.lua.org/download.html

"Lua also compiles cleanly as C++."

That quote may apparently be a lie in the current release, but at least the intent appears to be that this is supported (there is also discussion on the Lua mailing list to that effect, albeit very old discussion).

**Comment by milleniumbug** [comment]
@RaptorFactor So you may want to raise the issue to Lua devs, the fact you had to change the source code clearly means an unsupported configuration.

**Comment by TheCatPlusPlus** [comment]
Sure, the DLL frames will be able to pass the exception along, but the unwind will likely result in memory leaks and other kinds of fun state corruption. Don't throw exceptions in any kind of C library callbacks.

**Comment by milleniumbug** [comment]
Things not to do: don't try to remove comment and then try to make an extended response, because that will get you called a coward.

**Comment by RaptorFactor** [comment]
Okay so I've cleared up some of my confusion I think... I was coming off just testing some other languages like ChaiScript which contain native support for C++ exceptions, I stupidly applied that model to Lua without thinking. I forgot the whole "ANSI C" part, so compiling as C++ is not going to enable some magic handling of exceptions. The translation of exceptions to Lua errors is done at the sol2 level. Fine.

However... I still don't think I"m understanding the whole picture here. See my response to @TheCatPlusPlus .

@TheCatPlusPlus That's what I thought at first too, but if that's the case, what is the purpose of the exception handling in this code:
https://github.com/ThePhD/sol2/blob/b05ce972477cf0b9d96cba03cb512178faaf4746/sol/protected_function.hpp#L93

If I compile that using 'normal' settings (i.e. /EHsc), and Lua is compiled as C, won't the compiler assume that the extern "C" functions can't throw and hence attempts to throw will result in a call to std::terminate (or a crash... I forget whether it treats it as noexcept or __declspec(nothrow), which have different semantics iirc)?

@milleniumbug All I saw before I went to bed was that the message I responded to had disappeared. It's unfortunate that it seems to have been nothing more than a case of bad timing and so I apologize for that, but this is what happens when you delete entire messages instead of editing them (and doing so clearly - ninja edits are worse than deletions).

**Comment by TheCatPlusPlus** [comment]
@RaptorFactor Don't ask me, I don't know internals of either this or Lua. In general throwing across foreign frames is a bad idea, unless proven otherwise. Especially in 32-bit world, because 32-bit world is a goddamn mess. Dunno about terminate or extern C assumptions on MSVC part.

**Comment by ThePhD** [comment]
<_> So, uh. I need to fix the other bug related to `protected_function` first, to make sure it's not an artifact of that. Then I'll come back to this one and try and figure out what's going on. Also, school.

**Comment by RaptorFactor** [comment]
Thanks. Take your time, this isn't blocking anything for me, I'm just experimenting at the moment. I'll likely try and have a closer look on the weekend, but this is definitely not my area of expertise so no promises I'll have anything to show for it. ;)

**Comment by ThePhD** [comment]
So #238 might have been related to this and not entirely about exceptions working or not working, and it's fixed. Give it a shot: https://github.com/ThePhD/sol2/releases/tag/v2.14.9

**Comment by RaptorFactor** [comment]
Exactly the same error unfortunately (using 1be01b716ba1564aae0c3f18431efc71608d99b9).

**Comment by ThePhD** [comment]
Alrighty, so as everyone said Lua compiled as C++ is a bad idea. A really bad idea. But I was having some trouble with my protected_function implementation lately, and I needed to make sure that this wasn't my derp on a configuration that could potentially work anyways.

Lua is not meant to be compiled as C++, this is something everyone has told you. But the fun part here is what Lua _does_ when an exception happens when you compile the code as C++.

The stack unwinds, as it should. Lua even has try / catch code inside of itself in the C code to handle the exception. But this code-path is so untested and so broken it not only returns an impossible error code to the `protected_function_result` (`sol::call_status` enum value of -1, _which isn't defined by Lua_), but also absolutely skips performing any of the usual stack cleanup. That means: it does not remove the function from the stack, it does not clean up upvalues for that function, and it may/may not leave arguments scattered haphazardly all over the stack when it's done. None of the guarantees are kept, none of the invariants of the API call are preserved, and the stack is in an absolutely fucked state when it's done.

In other words, Lua completely shits the bed when compiled as C++ and things don't go _perfectly_.

Don't forcefully rewrite Lua ("make small edits to the source") and compile the whole library as C++. Exceptions can be transported even when compiling under C (you specify /EHsc and leave it alone). The code works with thrown exceptions as intended with LuaJIT and Lua 5.3/5.2/5.1, and we test all those configurations in travis-CI and on my own machine.

Remove the /TP flag, leave the `extern C { ... }`, and don't modify the code.

**Comment by RaptorFactor** [comment]
So everyone here keeps asserting that "Lua is not meant to be compiled as C++", and now clearly that is the case in practice (at least currently), yet once again I feel compelled to point out the Lua website and mailing list (and even the source code!) says the opposite... I'll go bug them instead in that regard, if only for reasons of my own morbid curiosity.

Sorry for the retarded bug report I guess...

**Comment by ThePhD** [comment]
It could very well be that it's meant to. I know the website and all say it's supposed to, and even the [mailing list](http://lua-users.org/lists/lua-l/2007-10/msg00478.html) like you said.

The problem is, reality is a lot more painful than the words.

As it stands, Lua 5.3.3 doesn't even compile cleanly in C++ (and not even with reduced warnings levels in VC++). That alone kind of hints that it's not really tested very often for C++. Furthermore, the only thing that will ignore destructors if you compile with C is `luaL_error()`, which uses setjmp / longjmp. If _you_ throw an error, the stack will unwind from your code, and Sol has trampolines in place to make sure everything stays clean.

IOW, you're just better of compiling as C and letting the safety guards I put in place for Sol work for you, and not let this weird implementation stuff mess you up.

**Comment by ThePhD** [comment]
Also, FWIW, it's not a retarded bug report. The Lua website and the mailing list and other Lua-curated parts (e.g., the lua-users wiki) all say it's fair game. Even Lua-Intf links to the lua-users wiki and uses that as the instructions for how to deal with C++ exceptions.

It's just that your question would probably end up better answered on the mailing list, than here, since in my case I have to abide fairly strictly by not just what the manual says, but how the runtime actually behaves.

**Comment by ThePhD** [comment]
Hey, I actually got to speak with one of the implementors (Roberto) of Lua. He told me this is probably a bug and they don't really vet the implementation for C++, even though it IS supposed to compile with it. I sent some fixes for compilation and a slimmed-down version of the issue for them to fix it up on the Lua Mailing List (I'll link it here when it a link to it gets serialized on the mailing list).
